test/inference.jl

```diff
@@ -491,3 +491,12 @@ tpara18457{I}(::Type{AbstractMyType18457{I}}) = I
 tpara18457{A<:AbstractMyType18457}(::Type{A}) = tpara18457(supertype(A))
 @test tpara18457(MyType18457{true}) === true
 
+fUnionAll{T}(::Type{T}) = Type{S} where S <: T
+@inferred fUnionAll(Real) == Type{T} where T <: Real
+@inferred fUnionAll(Rational{T} where T <: AbstractFloat) == Type{T} where T<:(Rational{S} where S <: AbstractFloat)
+
+fComplicatedUnionAll(::Type{T}) = Type{Tuple{S,rand() >= 0.5 ? Int : Float64}} where S <: T
+@test Base.parameter_upper_bound(fComplicatedUnionAll(Real), 1) == Real
+let x = fComplicatedUnionAll(Real).parameters[1].parameters[2]
```

Comment:
This test should be for `parameters[1] == Tuple{Real, Int} or Tuple{Real, Float64}`, since it is not true that `typeof(parameters[1]) === DataType`. (it it permitted to instead be `Tuple{T, Int} where T <: Real`)